natlib/svcd.c

```diff
@@ -185,3 +196,208 @@ static int svcd_init( lua_State *L )
+static int svcd_callback_handler(lua_State *L);
+static int anon_function(lua_State *L);
+
+static int svcd_callback_handler(lua_State *L){
+    int ivkd = (int) lua_tonumber(L, lua_upvalueindex(1));
+    printf("making callback function using ivkd = %i \n", ivkd);
+
+    resolve_table(L, "TIMEOUT");
+    int timeout = (int) lua_tonumber(L, -1);
+
+    resolve_table(L, "handlers"); // 7
+    lua_pushnumber(L, ivkd);
+    lua_gettable(L, -2);
+
+    int isnil = (int) lua_toboolean(L, -1);
+    if(isnil != 0){
+        lua_pushnumber(L, timeout);
+        lua_call(L, 1, 0);
+
+        resolve_table(L, "handlers"); // 7
+        lua_pushnumber(L, ivkd);
+        lua_pushnil(L);
+        lua_settable(L, -3);
+    }
+    return 0;
+}
+static int anon_function(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    printf("making anon function using ivkd = %i \n", ivkd);
+
+    lua_pushnumber(L, ivkd);
+    lua_pushcclosure(L, &svcd_callback_handler, 0);
+    return 1;
+}
+
+// write 
+static int svcd_write_invoke_later(lua_State *L){
+    int ivkd = (int)lua_tonumber(L, 1);
+    int timeout_ms = (int)lua_tonumber(L, 2);
+    lua_pushlightfunction(L, libstorm_os_invoke_later);
+    lua_pushnumber(L, timeout_ms * MILLISECOND_TICKS);
+    
+    printf("making custom anon funtion\n");
+    // lua_pushlightfunction(L, anon_function);
+    lua_getglobal(L, "anon_function");
+        lua_pushnumber(L, ivkd);
+        lua_call(L, 0, 1);
+    printf("made custom anon funtion\n");
+
+    lua_call(L, 2, 0);
+    return 0;
+}
+
+static int svcd_write( lua_State *L )
+{   
+    // REGISTER CLUA FUNCTIONS IN GLOBAL SPACE
+    // lua_pushcfunction(L, svcd_write_invoke_later);
+    // lua_setglobal(L, "svcd_write_invoke_later");
+    // lua_pushcfunction(L, anon_function);
+    // lua_setglobal(L, "anon_function");
+    // lua_pushcfunction(L, svcd_callback_handler);
+    // lua_setglobal(L, "svcd_callback_handler");
+    // END REGISTER
+
+    printf("Sexy write!\n");    
+    int numargs = lua_gettop(L); // 6
+  
+    // GET ALL THE ARGS
+    size_t g;
+    const char* targetip = (char*) lua_tolstring(L, 1, &g);
+    int svcid = (int)lua_tonumber(L, 2);
+    int attrid = (int)lua_tonumber(L, 3);
+    
+    size_t l;
+    const char* payload =  lua_tolstring(L, 4, &l);
+
+    int timeout_ms = (int)lua_tonumber(L, 5);
+    int* on_done = (int*) lua_tonumber(L, 6);
+     /* PARAMS */
+    printf("Params (%i) received ip:%s, svc: 0x%x, attr: 0x%x, payload:%s, ondone= 0x%x \n", numargs, targetip, svcid, attrid, payload, on_done);
+    
+    /* SET IVKID */
+    resolve_table(L, "ikvid"); //7
+    int ivkid = (int) lua_tonumber(L, -1);
+   
+    printf("Getting ikvid: %i\n", ivkid);
+  
+    ivkid =  ivkid + 1;
+    if( ivkid > 65535 ) { 
+        ivkid = 0;
+    }
+
+    lua_settop(L, numargs);
+
+    printf("Updating ikvid to %i\n", ivkid);
+    // setting ivkid
+    lua_getglobal(L, "SVCD");
+    lua_pushnumber(L, ivkid);
+    lua_setfield(L, -2, "ikvid");
```

Comment:
This chunk of code to increment ikvid exists twice in svcd.c, perhaps a they should both be replaced with a next_ikvid function
(Running With Scissors)